packages/langium/src/grammar/type-system/type-collector/types.ts

```diff
@@ -376,6 +388,114 @@ function pushReflectionInfo(node: CompositeGeneratorNode, name: string) {
     node.append('}', NL);
 }
 
+function pushDataTypeReflectionInfo(node: CompositeGeneratorNode, union: UnionType) {
+    const subTypes = Array.from(union.subTypes).map(e => e.name);
+    const strings = collectStringValuesFromDataType(union);
+    const regexes = collectRegexesFromDataType(union);
+    const returnString = createDataTypeCheckerFunctionReturnString(subTypes, strings, regexes);
+    node.append(`export function is${union.name}(item: string): item is ${union.name} {`, NL);
+    node.indent(body => {
+        body.append(returnString + ';', NL);
+    });
+    node.append('}', NL);
+}
+
+function createDataTypeCheckerFunctionReturnString(subTypes: string[], strings: string[], regexes: string[]): string {
+    let returnString = 'return ';
+    if (subTypes.length > 0) {
+        for (let i = 0; i < subTypes.length; i++) {
+            returnString += `is${subTypes[i]}(item)`;
+            if (i < subTypes.length - 1) {
+                returnString += ' || ';
+            }
+        }
+    }
+    if (subTypes.length > 0 && strings.length > 0) {
+        returnString += ' || ';
+    }
+    if (strings.length > 0) {
+        for (let i = 0; i < strings.length; i++) {
+            returnString += `item === '${strings[i]}'`;
+            if (i < strings.length - 1) {
+                returnString += ' || ';
+            }
+        }
+    }
+    if ((subTypes.length > 0 || strings.length >0 ) && regexes.length > 0) {
+        returnString += ' || ';
+    }
+    if (regexes.length > 0) {
+        for (let i = 0; i < regexes.length; i++) {
+            returnString += `/${regexes[i]}/.test(item)`;
+            if (i < regexes.length - 1) {
+                returnString +=' || ';
+            }
+        }
+    }
+    return returnString;
```

Comment:
This code seems rather lengthy. We should just collect the check/test code snippets once and then just join them together with `snippets.join(' || ')`.